Feature/js unit tests

llakie opened the pull request:
Add Node.js unit tests for all utilities and some critical core modules. (For an exact overview of the tested modules, please see single commits)
The tests can be executed using `npm test` or `yarn test` after you installed all dependencies using `yarn` (npm is discouraged, since we provide a yarn.lock file)
The HTML reports can be found under _/reports/coverage/index.html_ and as coberture.xml output.

Introduces new dev dependencies:

- eslint-plugin-jasmine MIT
- import-fresh MIT
- jasmine MIT
- mock-require MIT
- nyc ISC
- temp MIT

All dependencies are dynamically linked and installed on user side.

It will also include:

- FEATURE: Drop support of non MQTT5 brokers
- FEATURE: Make Platform Gateway support exports and named exports for Adapter instantiation
- FIX: Wrong match comparison in excludeOn list
- FEATURE: Introduce clear() method for featureMap
- FIX: Message stashing in MqttProtocolAdapter
- FIX: Subsequent rules depending on the same feature for different talents was not evaluated

which are issues found on the way integration the tests.

llakie committed:
Unit test setup for Node.js core modules

Signed-off-by: Kiefer Lars-Erich <Lars-Erich.Kiefer@de.bosch.com>

llakie committed:
Add unit tests for core.util.jsonQuery module

Signed-off-by: Kiefer Lars-Erich <Lars-Erich.Kiefer@de.bosch.com>

llakie committed:
Add unit tests for core.util.jsonModel

Signed-off-by: Kiefer Lars-Erich <Lars-Erich.Kiefer@de.bosch.com>

llakie committed:
Add unit tests for TalentIO Module

Signed-off-by: Kiefer Lars-Erich <Lars-Erich.Kiefer@de.bosch.com>

llakie committed:
Add no-coverage test script

Signed-off-by: Kiefer Lars-Erich <Lars-Erich.Kiefer@de.bosch.com>

llakie committed:
Add unit tests for ArrayPatternMatcher

Signed-off-by: Kiefer Lars-Erich <Lars-Erich.Kiefer@de.bosch.com>

llakie committed:
Add unit tests for clone util

Signed-off-by: Kiefer Lars-Erich <Lars-Erich.Kiefer@de.bosch.com>

llakie committed:
Add unit tests for equals utility

Signed-off-by: Kiefer Lars-Erich <Lars-Erich.Kiefer@de.bosch.com>

llakie committed:
Add unit tests for FeatureGraph

Signed-off-by: Kiefer Lars-Erich <Lars-Erich.Kiefer@de.bosch.com>

llakie committed:
Add unit tests for JsonModel

Signed-off-by: Kiefer Lars-Erich <Lars-Erich.Kiefer@de.bosch.com>

llakie committed:
Add unit tests for logger

Signed-off-by: Kiefer Lars-Erich <Lars-Erich.Kiefer@de.bosch.com>

llakie committed:
Add unit tests for FeatureMap

Signed-off-by: Kiefer Lars-Erich <Lars-Erich.Kiefer@de.bosch.com>

llakie committed:
Add test utilities for to prepare unit tests for MQTT driven modules

Signed-off-by: Kiefer Lars-Erich <Lars-Erich.Kiefer@de.bosch.com>

llakie committed:
Add unit tests for MQTTClient and MQTTProtocolAdapter

Signed-off-by: Kiefer Lars-Erich <Lars-Erich.Kiefer@de.bosch.com>

llakie committed:
Drop support for non-mqtt5 brokers

Signed-off-by: Kiefer Lars-Erich <Lars-Erich.Kiefer@de.bosch.com>

llakie committed:
Make modules and named exports usable as classes during ProtocolGateway adapter instantiation

Signed-off-by: Kiefer Lars-Erich <Lars-Erich.Kiefer@de.bosch.com>

llakie committed:
Fix wrong comparison for regex results in excludeOn strings

Signed-off-by: Kiefer Lars-Erich <Lars-Erich.Kiefer@de.bosch.com>

llakie committed:
Add clear method for featureMap to reduce the number of instantiations during unit tests

Signed-off-by: Kiefer Lars-Erich <Lars-Erich.Kiefer@de.bosch.com>

llakie committed:
Disable any log output during unit testing for MqttClient

Signed-off-by: Kiefer Lars-Erich <Lars-Erich.Kiefer@de.bosch.com>

llakie committed:
Add unit tests for rules module

Signed-off-by: Kiefer Lars-Erich <Lars-Erich.Kiefer@de.bosch.com>

llakie committed:
Update yarn.lock file

Signed-off-by: Kiefer Lars-Erich <Lars-Erich.Kiefer@de.bosch.com>

llakie committed:
Add unit tests for ProtocolGateway

Signed-off-by: Kiefer Lars-Erich <Lars-Erich.Kiefer@de.bosch.com>

llakie committed:
Add unit tests for InstanceManager

Signed-off-by: Kiefer Lars-Erich <Lars-Erich.Kiefer@de.bosch.com>

llakie committed:
Fix message stashing on publishing a message using the MqttClientAdapter

Signed-off-by: Kiefer Lars-Erich <Lars-Erich.Kiefer@de.bosch.com>

llakie committed:
Add unit tests for MetadataManager

Signed-off-by: Kiefer Lars-Erich <Lars-Erich.Kiefer@de.bosch.com>

llakie committed:
Add unit tests for Instance

Signed-off-by: Kiefer Lars-Erich <Lars-Erich.Kiefer@de.bosch.com>

llakie committed:
Add unit tests for timeseries rules

Signed-off-by: Kiefer Lars-Erich <Lars-Erich.Kiefer@de.bosch.com>

llakie committed:
Fix bug where subsequent rules depending on the same feature for different talents was not evaluated

Signed-off-by: Kiefer Lars-Erich <Lars-Erich.Kiefer@de.bosch.com>

llakie committed:
Prevent logging output during unit-tests

Signed-off-by: Kiefer Lars-Erich <Lars-Erich.Kiefer@de.bosch.com>

llakie committed:
Add unit tests for rules.nelson module

Signed-off-by: Kiefer Lars-Erich <Lars-Erich.Kiefer@de.bosch.com>

JochenKienzle commented on test/src/core/util/jsonModel.spec.js:
```diff
@@ -0,0 +1,53 @@
+        expect(model.get('bar.baz')).toBe('Hello World');
+     });
+
+     it('shold return the default value in case the path does not exist', () => {
```
typo should

JochenKienzle commented on test/src/core/util/jsonModel.spec.js:
```diff
@@ -0,0 +1,53 @@
+        model = new JsonModel(json);
+     });
+
+     it('shold return a given value', () => {
```
typo should

JochenKienzle commented on test/src/core/util/jsonModel.spec.js:
```diff
@@ -0,0 +1,48 @@
+        model = new JsonModel(json);
+    });
+
+    it('shold return a given value', () => {
```
typo shold

JochenKienzle commented on test/src/core/util/jsonModel.spec.js:
```diff
@@ -0,0 +1,48 @@
+        expect(model.get('bar.baz')).toBe('Hello World');
+    });
+
+    it('shold return the default value in case the path does not exist', () => {
```
typo shold

JochenKienzle commented on test/src/core/util/jsonModel.spec.js:
```diff
@@ -0,0 +1,48 @@
+        }).toThrow();
+    });
+
+    it('shold return a submodel at a given path', () => {
```
typo shold (check others as well ;-)

JochenKienzle commented on test/src/core/util/jsonModel.spec.js:
```diff
@@ -0,0 +1,48 @@
+        expect(() => model.get('bar.bar')).toThrow();
+    });
+
+    it('it should validate a model, with a given validator function', () => {
```
consistency: "it should" vs "should" (check others as well)

JochenKienzle commented on test/src/core/util/talentIO.spec.js:
```diff
@@ -0,0 +1,143 @@
+describe('core.util.talentIO', () => {
+    const ev = require('../../../resources/talentIO.event.json');
+
+    describe('TalentInput', () => {
```
These tests seems to only tests for good cases but no error case (e.g. invalid / broken json)

llakie replied:
Since the json comes from our platform, it's ensured, that only well formatted JSON will be taken. Since jsonModel and jsonQuery are used to retrieve all the stuff, which have built-in error handling, it's currently not necessary to test the "unhappy" path. We can add this in the future.

JochenKienzle replied:
I agree with you that the platform has to provide a correct json but I see TalentIO still dedicated from it and TalentIO (as a central point) should not result in an undefined behaviour if something gets broken. I think the point is now how the TalentIO will handle such cases... I assume that it is already throwing an error/exception (Which could be tested here) or it is something which has to be added as a feature later on.

llakie replied:
> I assume that it is already throwing an error/exception (Which could be tested here) or it is something which has to be added as a feature later on.

Yes, it will throw a self explaining exception, if anything goes wrong

JochenKienzle replied:
Ah sorry I was not aware that jsonModel & jsonQuery is part of TalentIO and are repsonsible for that (and has been tested there already)

JochenKienzle commented on test/src/core/util/featureGraph.spec.js:
```diff
@@ -0,0 +1,89 @@
+
```
unnecessary newlines

JochenKienzle commented on test/src/core/util/featureGraph.spec.js:
```diff
@@ -0,0 +1,89 @@
+        expect(fg.ids).toEqual(['1', '2', '3']);
+    });
+
+    it('should also being able to handle wildcard dependency as from dependency', () => {
```
(this comment applies to other test cases)
I think the word 'also' is unnecessary 
Or shall it indicate that this case is not really intended to be working (optional)?

llakie replied:
I will change the message. It is an artifact from having written a test for a fixed feature. Therefore the word "also" snugged in.

JochenKienzle commented on test/src/core/util/featureGraph.spec.js:
```diff
@@ -0,0 +1,89 @@
+        expect(fg.containsCycles()).toBeFalsy();
+    });
+
+    it('should always detect a cycle, if any-wildcard is used', () => {
```
I think 'always' is also a little bit 'too much' because you are only testing a single cycle case. (I assume there could be much more complex ones and under different conditions)

llakie replied:
The test does, what the spec implies. If you use a `*.*` wildcard, you will always end up with a cycle.

JochenKienzle replied:
I got it. I understand it wrongly at first place.

JochenKienzle commented on src/core/util/logger.js:
```diff
@@ -13,34 +13,36 @@ const jsonata = require('jsonata');
 
-    verbose(message, ctx) {
-        this.__log(message, Logger.__LOG_LEVEL.VERBOSE, ctx);
+    verbose(message, ctx, now) {
```
I think 'now' has been introduced to provide tests for that but from the perspective of a feature it makes not much sense for me (if it is used it doesn't improve anything and it opens the possibility to create logs "out of sync")

We could think about having a "timestamp for the related event" .. e.g. if an error is logged with a delay and we want to show the timestamp of the exact event (but this could of course be simply part of the message).

I have seen that you can change the logfunction --> can this not be used to hide / avoid "now"?

llakie replied:
You can actually put a logging facade in front of the logger (or make the current Logger class a delegate to the acutal implementation), which is used by the user and still have the actual logger implementation testable by exposing the current timestamp. This might take a while, since we are exporting the logger module directly and not via a named export. To implement a testable facade, we would have to export both, the facade and the actual implementation at the same time and update the imports using object destructuring. like `const { Logger } = require('...')`

> I have seen that you can change the logfunction --> can this not be used to hide / avoid "now"?

The log function does not help, since the timestamp is passed in as a formatted parameter into this function

JochenKienzle replied:
So, it would be something for a later step. Would be ok for me

JochenKienzle commented on test/src/core/util/logger.spec.js:
```diff
@@ -0,0 +1,134 @@
+        expect(console.error.calls.mostRecent().args[0]).toMatch(/An error \[Error: .+\]$/g);
+    });
+
+    it('should render an log message console.log', () => {
```
I'm missing a test with an invalid object / undefined as message, because I remember that this will lead to a 'strange' error (input.replace is not a function). Maybe this has been fixed already, but a test for such cases would be nice (to avoid to print some strange (and maybe big) things.

llakie replied:
I will add this test for sure, since it lead to a platform crash

llakie replied:
Added new function __serializeMessage to logger, which will take care of trying to serialize a given message/object/number... If it fails, placeholders can be specified when the input is `null` or `undefined` `default = '<empty>'` or when it's an unserializable input `default = '<unserializable>'` and added tests for invalid input parameters.

JochenKienzle commented on test/src/core/util/featureMap.spec.js:
```diff
@@ -0,0 +1,56 @@
+
```
unnecessary newlines

JochenKienzle commented on test/src/core/util/featureMap.spec.js:
```diff
@@ -0,0 +1,56 @@
+        expect(fm.get('default', 'test', '4711').matches).toBe(1);
+    });
+
+    it('should store the metadata on an type.feature level', () => {
```
typo "an"

JochenKienzle commented on test/helpers/mock/mqtt.mock.js:
```diff
@@ -0,0 +1,96 @@
+const {
```
Missing license header

llakie replied:
Done

JochenKienzle commented on test/helpers/null.logger.js:
```diff
@@ -0,0 +1,9 @@
+const Logger = require('../../src/core/util/logger');
```
Missing license header

JochenKienzle commented on test/helpers/protocolGateway.forcewait.js:
```diff
@@ -0,0 +1,26 @@
+function forceWaitOnPublish(pg) {
```
Missing license header

llakie replied:
It's there - this might be the refreshing problem again. You can check the latest version of the branch

See: https://github.com/GENIVI/iot-event-analytics/pull/53/files#diff-099043d3fe2313c97ee5ad499e960936e2ca453f82d52c190ad5ceea17c02172

argerus commented on src/core/routing.js:
```diff
@@ -94,13 +94,13 @@ module.exports = class Routing {
-
                 await this.pg.publishJson(Talent.getTalentTopic(id, ev.value.$tsuffix), Object.assign(
+                    {},
                     ev,
```
Maybe `ev` should be [frozen](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/freeze) or otherwise be made immutable in order to prevent these type of bugs.

Also, do we have a new unit test catching this bug now?

JochenKienzle replied:
I don't see any routing.spec file so I assume they are missing. (Is it planned for later?)

llakie replied:
@JochenKienzle Yes, routing spec is coming up. @argerus Good point with Object freezing to make ev being immutable in the routing component is a good idea (Where the same event is used multiple times). In other components like encoding, where an event is processed once, we are currently enriching this event by several properties. So freezing in this very case avoids this nasty error in the future in the routing component.

llakie replied:
I will implement that in another PR

JochenKienzle reviewed (changes requested):
Overall it looks good and most of the comments are related to consistency and some typos except:

- missing tests for routing -> Is it planned for later?
- I am missing tests for invalid cases (invalid json, undefined parameters,...) but this could be done later
- introduction of "now" in logger
- Some license headers a missing (but I'm not sure if they are necessary on mocks)

llakie commented:
Thank you for the review. I will work on implementing the required changes.

llakie commented:
> missing tests for routing -> Is it planned for later?

Yes, it's planned but not yet done

> I am missing tests for invalid cases (invalid json, undefined parameters,...) but this could be done later

Implemented for the logger now

> introduction of "now" in logger

Will fix this at a later stage, since now is an optional parameter and will default to Date.now() if not given. We can still think of this facade as I pointed out earlier

> Some license headers a missing (but I'm not sure if they are necessary on mocks)

Were all added

llakie committed:
Added requested features from review

- Implement testcases for invalid logger messages. Make the logger implementation fail-proof
- Fixed typos
- Add license headers where they were missing

Signed-off-by: Kiefer Lars-Erich <Lars-Erich.Kiefer@de.bosch.com>

JochenKienzle commented on test/src/core/protocolGateway.spec.js:
```diff
@@ -0,0 +1,228 @@
+        return new ProtocolGateway(importFresh('../../resources/protocolGateway.config.json'), displayName, usePlatformProtocolOnly);
+    }
+
+    it('It should publish messages to all defined adapters', async () => {
```
Consistency: This file still uses these 'It should ...' pattern

llakie replied:
Sorry, I just have overseen that. Fixed it

JochenKienzle reviewed (approved):
Looks good for me. 👍
